{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3548e26e-bc2f-4efe-87eb-8b9e1d24e9fc",
   "metadata": {},
   "source": [
    "# Sparks\n",
    "## Multi-modal, multi-agent AI model system capable of independently conducting research by formulating hypotheses, performing experiments, and adapting its strategy\n",
    "A. Ghafarollahi, M.J. Buehler*\n",
    "\n",
    "Massachusetts Institute of Technology\n",
    "\n",
    "*mbuehler@MIT.EDU"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ac9218-4b71-49d9-a36d-80f9bae5e161",
   "metadata": {},
   "source": [
    "# OpenAI API Key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9ba4e2d-41a0-4d54-8f99-fe2702bbef13",
   "metadata": {},
   "outputs": [],
   "source": [
    "your_openai_api_key = 'sk-'\n",
    "os.environ['OPENAI_API_KEY'] = your_openai_api_key"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b41d7d4-f561-4659-9559-a4098428a0e7",
   "metadata": {},
   "source": [
    "# Query, Tools, Constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b817dd29-3434-4b85-af6d-2749edef3541",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory exists.\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "import os\n",
    "\n",
    "\n",
    "base_dir = 'conversations'\n",
    "\n",
    "try:\n",
    "    os.mkdir(base_dir)\n",
    "except:\n",
    "    print('Directory exists.')\n",
    "\n",
    "tools =  '''\n",
    "{\n",
    "  'type': 'Python function',\n",
    "  'function': {\n",
    "    'name': 'calculate_force_from_seq',\n",
    "    'description': (\n",
    "      'This function takes a protein sequence in one-letter amino acid code, '\n",
    "      'performs the required analysis or simulation, and returns the '\n",
    "      'non-dimensional protein force (defined as the maximum force along the protein unfolding force-extension curve) as a float.'\n",
    "      'The function might lead to None results. So, the output should always be checked.'\n",
    "    ),\n",
    "    'parameters': {\n",
    "      'type': 'object',\n",
    "      'properties': {\n",
    "        'sequence': {\n",
    "          'type': 'string',\n",
    "          'description': (\n",
    "            'Protein sequence represented using standard one-letter amino acid codes '\n",
    "            '(e.g., \"MKTFFV...\").'\n",
    "          )\n",
    "        }\n",
    "      },\n",
    "      'required': ['sequence']\n",
    "    }\n",
    "  }\n",
    "},\n",
    "{\n",
    "  'type': 'Python function',\n",
    "  'function': {\n",
    "    'name': 'calculate_energy_from_seq',\n",
    "    'description': (\n",
    "      'This function takes a protein sequence in one-letter amino acid code, '\n",
    "      'performs the required analysis or simulation, and returns the '\n",
    "      'non-dimensional protein energy (defined as the area under the protein unfolding force-extension curve) as a float.'\n",
    "      'The function might lead to None results. So, the output should always be checked.'\n",
    "    ),\n",
    "    'parameters': {\n",
    "      'type': 'object',\n",
    "      'properties': {\n",
    "        'sequence': {\n",
    "          'type': 'string',\n",
    "          'description': (\n",
    "            'Protein sequence represented using standard one-letter amino acid codes '\n",
    "            '(e.g., \"MKTFFV...\").'\n",
    "          )\n",
    "        }\n",
    "      },\n",
    "      'required': ['sequence']\n",
    "    }\n",
    "  }\n",
    "},\n",
    "  {\n",
    "  'type': 'Python function',\n",
    "  'function': {\n",
    "    'name': 'design_protein_from_length',\n",
    "    'description': (\n",
    "      'This function creates random protein sequences of a given length. '\n",
    "      'It returns the amino acid sequence (string format) of the generated protein.'\n",
    "    ),\n",
    "    'parameters': {\n",
    "      'type': 'object',\n",
    "      'properties': {\n",
    "        'length': {\n",
    "          'type': 'integer',\n",
    "          'description': 'Desired number of amino acids in the protein sequence'\n",
    "        }\n",
    "      },\n",
    "      'required': ['length']\n",
    "    }\n",
    "  }\n",
    "},\n",
    "{\n",
    "  'type': 'Python function',\n",
    "  'function': {\n",
    "    'name': 'design_protein_from_CATH',\n",
    "    'description': (\n",
    "      'This function creates random alpha-helix, beta-sheet, or mixed protein sequences of a given length. '\n",
    "      'It returns the amino acid sequence (string format) of the generated protein. '\n",
    "      'The function can be used to create proteins biased toward alpha-helix (CATH=1), beta-sheet (CATH=2), or mixed structures (CATH=3). '\n",
    "      'However, the output may not fully reflect the desired structural class, especially for short sequences.'\n",
    "      'To ensure the proteins follow the desired structure, the secondary streucture of folded sequences should be analyzed.'\n",
    "    ),\n",
    "    'parameters': {\n",
    "      'type': 'object',\n",
    "      'properties': {\n",
    "        'length': {\n",
    "          'type': 'integer',\n",
    "          'description': 'Desired number of amino acids in the protein sequence'\n",
    "        },\n",
    "        'cath': {\n",
    "          'type': 'string',\n",
    "          'description': 'Structure bias: \"1\" for alpha-helix, \"2\" for beta-sheet, \"3\" for mixed'\n",
    "        }\n",
    "      },\n",
    "      'required': ['length', 'cath']\n",
    "    }\n",
    "  }\n",
    "},\n",
    "{\n",
    "  'type': 'Python function',\n",
    "  'function': {\n",
    "    'name': 'analyze_protein_structure',\n",
    "    'description': (\n",
    "      'This function returns the secondary structure percentage content of a protein from a PDB file. '\n",
    "      'The output is a dictionary with 8-class DSSP secondary structure elements as keys: '\n",
    "      '{\"H\", \"B\", \"E\", \"G\", \"I\", \"T\", \"S\", \"P\", \"-\"}. Each key maps to the corresponding percentage value (float).'\n",
    "    ),\n",
    "    'parameters': {\n",
    "      'type': 'object',\n",
    "      'properties': {\n",
    "        'protein_struct': {\n",
    "          'type': 'string',\n",
    "          'description': 'Protein structure file in PDB format'\n",
    "        }\n",
    "      },\n",
    "      'required': ['protein_struct']\n",
    "    }\n",
    "  }\n",
    "},\n",
    "{\n",
    "  'type': 'Python function',\n",
    "  'function': {\n",
    "    'name': 'fold_protein',\n",
    "    'description': (\n",
    "      'This function takes a protein amino acid sequence, folds it and creates the protein PDB file. Returns the PDB file name (string format)'\n",
    "      'of the resulting 3D folded structure. The sequence should be in FASTA format and contain only valid '\n",
    "      'amino acid characters (no special symbols like \"-\").'\n",
    "    ),\n",
    "    'parameters': {\n",
    "      'type': 'object',\n",
    "      'properties': {\n",
    "        'sequence': {\n",
    "          'type': 'string',\n",
    "          'description': 'Protein sequence in FASTA format (only amino acid characters, no special symbols).'\n",
    "        },\n",
    "        'name': {\n",
    "          'type': 'string',\n",
    "          'description': 'Name to be used for saving the folded protein structure.'\n",
    "        }\n",
    "      },\n",
    "      'required': ['sequence', 'name']\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "     '''\n",
    "\n",
    "constraints = '''\n",
    "1- During the research, assume protein lengths to be equal or between 30 and 80.\n",
    "2- Consider protein sequences in intervals of 10.\n",
    "3- For computational efficiency, restrict the total number of samples per protein length to 10.\n",
    "'''\n",
    "\n",
    "constraints = json.dumps(constraints, indent=4)\n",
    "\n",
    "total_followup_rounds = 3\n",
    "query = 'Discover a novel principle in short protein peptides.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2556fdf8-ecfa-47d4-82bf-9732ce64978b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Sparks_functions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "544b74e1-724a-4dec-a493-cfd217679a8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Idea generation\n",
    "generate_idea(query, tools, constraints)\n",
    "# Idea testing\n",
    "test_idea(query, tools, constraints)\n",
    "initial_implementation()\n",
    "# Refinement\n",
    "last_exp_round = refine_approach(query, constraints, total_followup_rounds)\n",
    "# Create plots\n",
    "create_reason_plots(query, last_exp_round)\n",
    "# Documentation\n",
    "create_introduction(query)\n",
    "create_methods(query, tools, constraints, last_exp_round)\n",
    "create_results()\n",
    "create_conclusion(query)\n",
    "create_outlook(query, tools)\n",
    "create_pdf()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
